src/Sarif/Visitors/InsertOptionalDataVisitor.cs

```diff
@@ -163,14 +163,15 @@ public override Message VisitMessage(Message node)
             if ((node.Text == null || _dataToInsert.HasFlag(OptionallyEmittedData.OverwriteExistingData)) &&
                 _dataToInsert.HasFlag(OptionallyEmittedData.FlattenedMessages))
             {
-                Rule rule = null;
                 string formatString = null;
-                if (_ruleId != null &&
-                    _run.Resources?.Rules.TryGetValue(_ruleId, out rule) == true)
+                Rule rule = _ruleIndex != -1 ? _run.Resources.Rules[_ruleIndex] : null;
+
+                if (rule != null && rule.MessageStrings.TryGetValue(node.MessageId, out formatString))
```

Comment:
>.T [](start = 55, length = 2)

Careful of `null`: `rule.MessageStrings?.TryGetValue`.

Reply:
nice catch. what a fantastic corner case. adding unit test.

---
In reply to: [249157825](https://github.com/Microsoft/sarif-sdk/pull/1223#discussion_r249157825) [](ancestors = 249157825)

Reply:
I don't think this is a sufficient fix. It assumes that a message id that points into a rule's message strings will never be the same as a message id that points into `run.resources.messageStrings`. It's easy to imagine message ids like `"default"` or `"success"` that might occur in both places (and even if it weren't easy to imagine, it might still happen).

What do you think of the fix I suggested above (reset `_ruleIndex` to -1 as soon as you use it)?

---
In reply to: [249194811](https://github.com/Microsoft/sarif-sdk/pull/1223#discussion_r249194811) [](ancestors = 249194811,249157825)

Reply:
See my comment in the unit tests, where I describe the failure case I'm worried about.

---
In reply to: [249229167](https://github.com/Microsoft/sarif-sdk/pull/1223#discussion_r249229167) [](ancestors = 249229167,249194811,249157825)

Reply:
I don't understand your suggestion. When we are visiting a result with a corresponding resultIndex, we retain it. If we reset this to -1 on first use, as you suggest, no other message visiting in context of the result would have access to this data. This seems like entirely the wrong thing to so. Remember: stacks, codeflows, graphs, fixes, etc. might dip into the rule resources to get format string.

We explicitly designed string look-up to refer to the rule first, then fall back to the global. Maybe your objections will be clearer to me when i look at your unit test remarks.

[https://github.com/oasis-tcs/sarif-spec/issues/216](https://github.com/oasis-tcs/sarif-spec/issues/216)
"in the presence of a non-null message.messageId, consult resources.rules by rule id, to attempt to locate the string. if not present, fall back to resources for look-up."

---
In reply to: [249229167](https://github.com/Microsoft/sarif-sdk/pull/1223#discussion_r249229167) [](ancestors = 249229167,249194811,249157825)